Convert Datastore to GAPIC

quartzmo opened the pull request:
This PR converts all Datastore requests to the GAX client introduced in #884.

There is a breaking change: The `retries` parameter is no longer supported (although method signatures have not changed.) See #848 and #849 for proposals to add new configuration options for retry behavior.

Also, Datastore is one of the three services that still requires expanded acceptance testing (#759). I could add the needed tests as part of this PR if desired.

quartzmo commented:
This PR should also remove files that are no longer needed for backoff, to close #893.

blowmage commented:
Not all gems are converted to GAX, so the backoff code should stay until they are all converted.

quartzmo commented:
Oh right, I mistakenly thought it wasn't shared. Of course.

quartzmo commented:
The [fourth commit](https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/903/commits/87178674eaf1914001e794acbc0d8d7c15c0503b) on this PR removes `retries` (no longer supported) from the factory methods, and adds a new `client_config` parameter accepting a hash of `google-gax` [`client_config`](http://www.rubydoc.info/gems/google-gax/0.5.1/Google%2FGax.construct_settings) values. Our intention with this change is to greatly increase client control over retry, timeout, and error handling behaviors, addressing issues such as #808.

Here is the default  [`datastore_client_config.json`](https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-datastore/lib/google/cloud/datastore/v1/datastore_client_config.json), which can be overridden via the new parameter:

```js
{
  "interfaces": {
    "google.datastore.v1.Datastore": {
      "retry_codes": {
        "retry_codes_def": {
          "idempotent": [
            "DEADLINE_EXCEEDED",
            "UNAVAILABLE"
          ],
          "non_idempotent": []
        }
      },
      "retry_params": {
        "default": {
          "initial_retry_delay_millis": 100,
          "retry_delay_multiplier": 1.3,
          "max_retry_delay_millis": 60000,
          "initial_rpc_timeout_millis": 60000,
          "rpc_timeout_multiplier": 1.0,
          "max_rpc_timeout_millis": 60000,
          "total_timeout_millis": 600000
        }
      },
      "methods": {
        "Lookup": {
          "timeout_millis": 60000,
          "retry_codes_name": "idempotent",
          "retry_params_name": "default"
        },
        "RunQuery": {
          "timeout_millis": 60000,
          "retry_codes_name": "idempotent",
          "retry_params_name": "default"
        },
        "BeginTransaction": {
          "timeout_millis": 60000,
          "retry_codes_name": "non_idempotent",
          "retry_params_name": "default"
        },
        "Commit": {
          "timeout_millis": 60000,
          "retry_codes_name": "non_idempotent",
          "retry_params_name": "default"
        },
        "Rollback": {
          "timeout_millis": 60000,
          "retry_codes_name": "non_idempotent",
          "retry_params_name": "default"
        },
        "AllocateIds": {
          "timeout_millis": 60000,
          "retry_codes_name": "non_idempotent",
          "retry_params_name": "default"
        }
      }
    }
  }
}
``` 

As you can see, there are some interesting fields, such as `retry_codes`, which suggests the ability to configure retry logic based on error status codes. At this time there is not much documentation for the [`client_config`](http://www.rubydoc.info/gems/google-gax/0.5.1/Google%2FGax.construct_settings) fields, but @jmuk from `google-gax` is working with us and should be able to answer questions.

Please note that this leaks the GAPIC implementation to the `google-cloud-datastore` public API. If `google-gax` changes its implementation, this may result in a change to the `google-cloud-datastore` public API.

@timanovsky @bmclean Any feedback on this proposal? We hope to merge and release this PR next week.

bmclean commented on google-cloud-datastore/lib/google-cloud-datastore.rb:
```diff
@@ -38,8 +38,9 @@ module Cloud
-    # @param [Integer] retries This option is not currently supported.
     # @param [Integer] timeout Default timeout to use in requests. Optional.
+    # @param [Hash] client_config A hash of values to override the default
+    #   behavior of the API client. See Google::Gax::CallSettings. Optional.
```
It isn't obvious that the parameters are in Google::Gax::BackoffSettings.

bmclean commented:
@quartzmo Please correct me if I have any of this wrong:
For a datastore Lookup, since it has a retry_codes_name defined of "idempotent", if an UNAVAILABLE error was raised, it would retry using the default params.
For a datastore Commit,  since it has a retry_codes_name defined of "non_idempotent", any error raised would not be retried.

blowmage commented:
@bmclean I'll take a swing at answering your question. Yes, your understanding is correct. That is the default behavior. We asked several times which endpoints we should retry and which ones we should not. But we never got an answer so we did a blanket retry on them all. I believe that this new behavior is correct.

Of course, the good news is that you can now override how each endpoint is treated and use whatever behavior you wish. :)

blowmage commented on google-cloud-datastore/lib/google-cloud-datastore.rb:
```diff
@@ -108,12 +108,12 @@ def datastore scope: nil, retries: nil, timeout: nil
     #     t["priority"] = 4
     #     t["description"] = "Learn Cloud Datastore"
-    #   end
+    #   end                                                                  ``
```
What is this? 😆

blowmage commented on google-cloud-datastore/lib/google/cloud/datastore/service.rb:
```diff
@@ -170,6 +141,13 @@ def generate_read_options consistency, transaction
+        def execute
+          require "grpc" # Ensure GRPC is loaded before rescuing exception
+          yield
+        rescue GRPC::BadStatus => e
```
Need to rescue `Google::Gax::GaxError` here.

bmclean commented:
@blowmage What do you mean by "blanket retry on them all"?

blowmage commented:
Correct. Not all errors, and not all endpoints should be retried. Of course, you are free to override the default configuration and make it behave as you want.

bmclean commented:
Thanks. I agree that the proposed default config is correct. For Lookup and RunQuery, if datastore returns either a DEADLINE_EXCEEDED or UNAVAILABLE error the request is retried as per the retry configuration. For all other endpoints, the default is not to invoke the retry logic.

blowmage commented:
This new configuration comes from Google. It has been hard for those of us outside of Google to get this information, so we are happy to finally get this level of guidance. :smile:

quartzmo commented:
@blowmage I believe I've updated this PR with everything it needs to match the merged GAPIC conversions for Vision and Pub/Sub. PTAL!

quartzmo commented:
oops wrong comment button
